src/librustdoc/html/markdown.rs

```diff
@@ -1124,56 +1128,70 @@ crate fn plain_text_summary(md: &str) -> String {
     s
 }
 
-crate fn markdown_links(md: &str) -> Vec<(String, Option<Range<usize>>)> {
+crate fn markdown_links(md: &str) -> Vec<(String, Range<usize>)> {
     if md.is_empty() {
         return vec![];
     }
 
-    let mut links = vec![];
-    let mut shortcut_links = vec![];
-
-    {
-        let locate = |s: &str| unsafe {
-            let s_start = s.as_ptr();
-            let s_end = s_start.add(s.len());
-            let md_start = md.as_ptr();
-            let md_end = md_start.add(md.len());
-            if md_start <= s_start && s_end <= md_end {
-                let start = s_start.offset_from(md_start) as usize;
-                let end = s_end.offset_from(md_start) as usize;
-                Some(start..end)
-            } else {
-                None
+    let links = RefCell::new(vec![]);
+
+    let locate = |s: &str, fallback: Range<usize>| unsafe {
+        let s_start = s.as_ptr();
+        let s_end = s_start.add(s.len());
+        let md_start = md.as_ptr();
+        let md_end = md_start.add(md.len());
+        if md_start <= s_start && s_end <= md_end {
+            let start = s_start.offset_from(md_start) as usize;
+            let end = s_end.offset_from(md_start) as usize;
+            start..end
+        } else {
+            fallback
+        }
+    };
+
+    let span_for_link = |link: &CowStr<'_>, span: Range<usize>| {
+        match link {
+            // Ranges provided for Borrowed strings may differ from where the actual string comes
+            // from. Since the &str points into `md`, we can calculate its actual location.
+            CowStr::Borrowed(s) => locate(s, span),
```

Comment:
Can you explain why only `Borrowed` variants have the wrong span? I would expect instead for it to be wrong for all spans, but we'd only be able to use `locate` on Borrowed variants because otherwise it points to a different allocation.

That brings up a couple other questions I have:
1. Why does `locate` only work for things within the same allocation? Can we improve it to work regardless?
2. When would pulldown give us a `Boxed` or `Inlined` span?

I think this PR would be good to land regardless, but the comment seems off, we should at least fix that before landing.

Reply:
Let's get the easy ones out of the way first:

 * The comment tries to hint at the current problem: we get a `Borrowed` string that contains the destination of a reference-style link, and a range that points at the link usage itself.

 * I'm not sure about `Inlined`, but it seems like it's a small string optimization for the `Boxed` variant. It exists to avoid allocating on `clone` if the boxed string is short enough.


As far as I can see, Pulldown provides a `Borrowed` if the string directly comes from the source. If any transformation is necessary (e.g. [unescaping](https://github.com/raphlinus/pulldown-cmark/blob/d4bf0872b14f68c1afedee918710fb401e3e6e9a/src/scanners.rs#L960) used when [scanning ref defs](https://github.com/raphlinus/pulldown-cmark/blob/d4bf0872b14f68c1afedee918710fb401e3e6e9a/src/parse.rs#L1408) or such), Pulldown builds and hands out a `Boxed` string.

The issue is not that `Borrowed` strings can have an incorrect span, it's that for those, we can actually restore the span we need. For any other variant, it's reasonable (I think) to assume that the string contained in the variant is some sort of derivative of the original source, and can not be located by either pointer arithmetic or string search. (Just as I write this, if I'm correct this makes `rfind`-ing on the `Boxed` variant kinda pointless, because the string will not be an exact match anyway. Otherwise, Pulldown would have handed out a `Borrowed`).

Hope this makes sense.

Reply:
> The issue is not that Borrowed strings can have an incorrect span, it's that for those, we can actually restore the span we need. 

Right, that's what I expected. Can you say that in the comment? Right now it makes it sound like the range will *only* be different for Borrowed, which isn't true.

Can you take a look at 1 to see if it's possible to get rid of the unsafe? (and hopefully make this work for `Boxed` variants too, but after your later comments that might not be possible.)

Reply:
I'm not completely certain what you're referring to as "1". Getting rid of the unsafe is as easy as casting the pointer to `usize` but that's just masking the problem. Otherwise, I would not change `locate` if there's a chance for a pulldown update that allows us to remove it entirely.

Reply:
Ok, r=me with the comment fixed then, and hopefully we can get rid of locate in a pulldown update soon.

Reply:
I found where my expectation break down: LinkReplacer can give us `Borrow` variants that borrow from something other than `md`. I made a note of this.